[Gekidou] Groups are included in the @metion auto-complete

shazm opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary

I think @larkox had done a lot of work for this already; I've just fixed up to make the correct call to get groups.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-43985

#### Checklist
<!--
Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.
-->
- [ ] Added or updated unit tests (required for all new features)
- [x] Has UI changes
- [ ] Includes text changes and localization file updates
- [ ] Have tested against the 5 core themes to ensure consistency between them.

#### Device Information
This PR was tested on: <!-- Device name(s), OS version(s) -->

#### Screenshots


https://user-images.githubusercontent.com/456511/167765379-3e2bb87d-6873-46ea-9e02-b2231f821c30.mp4



#### Release Note
<!--
Add a release note for each of the following conditions:

* New features and improvements, including behavioural changes, UI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense. Newlines are stripped.

Example:

```release-note
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```release-note
NONE
```
-->

```release-note

```

shazm commented on app/actions/remote/groups.ts:
```diff
@@ -4,13 +4,13 @@
     }
 
-    return client.getAllGroupsAssociatedToChannel(channelId, true);
+    return client.getGroups(query);
```
Yeah, in the webApp it lives as `searchGroups( ...`, but it's still a `GET` request to the root `/groups?q=...` endpoint with a slightly different query param.

Server code is [here](https://github.com/mattermost/mattermost-server/blob/856f9b11af5f12856969749934df56f1d4fcbdc3/api4/group.go#L96) and lightly out of date docs [here](https://api.mattermost.com/#operation/GetGroups)

mkraft replied:
Yeah, the webapp makes a GET request to `/api/v4/groups?q=&filter_allow_reference=true&page=0&per_page=60&include_member_count=true`. Note that the rest request here in the mobile app omits `include_member_count=true`, but I think that makes sense because we're not displaying the group member count. I do think we should in the future, but that's out of scope of the work we're doing here.

mkraft reviewed (commented):
Were the tests existing from the previous work by @larkox or does this need net new tests or test updates?

I'm noticing some UI placement issues for the groups icon and the loading animations, should we fix in a separate PR?

<img width="464" alt="Screen Shot 2022-05-11 at 7 22 59 PM" src="https://user-images.githubusercontent.com/1149597/167963290-6fbb37cb-a7fc-413d-9a5f-810169a1adf2.png">

shazm commented:
Updated to merge into Gekidou with schema dependency.

larkox commented:
> Were the tests existing from the previous work by @larkox or does this need net new tests or test updates?

Right now there are not unit tests for these parts of the code. Sooner or later they have to be added, but that is the main reason they haven't been updated.

Regarding final look, since we didn't have groups properly working at that time, I set up the skeleton, but was not able to do proper testing, so it makes sense there are still issues with the UI. It would be great if @shazm could test that and apply any needed fix in this PR.

enahum reviewed (commented):
pending the UX fixes

shazm commented:
> Were the tests existing from the previous work by @larkox or does this need net new tests or test updates?
> 
> I'm noticing some UI placement issues for the groups icon and the autosuggest loading animations, should we fix in a separate PR?
> 
> <img alt="Screen Shot 2022-05-11 at 7 22 59 PM" width="464" src="https://user-images.githubusercontent.com/1149597/167963290-6fbb37cb-a7fc-413d-9a5f-810169a1adf2.png">

I'm fixing in this one

shazm commented:
<img width="520" alt="Screen Shot 2022-05-16 at 11 08 43 am" src="https://user-images.githubusercontent.com/456511/168504070-2ab9cddd-f009-4417-b4f0-3fac12ff93e4.png">
<img width="648" alt="Screen Shot 2022-05-16 at 11 17 00 am" src="https://user-images.githubusercontent.com/456511/168504076-98e8b597-2e69-407b-8c19-dbdd89118d39.png">

Rebased off latest gekidou and updated styling.

shazm commented:
<img width="1037" alt="Screen Shot 2022-05-17 at 3 07 11 pm" src="https://user-images.githubusercontent.com/456511/168733574-c75ceb00-cd0a-4010-9e37-335b1c6776ff.png">

`ActivityIndicator` styles also fixed.

larkox commented on app/components/autocomplete/at_mention_group/at_mention_group.tsx:
```diff
@@ -16,34 +16,41 @@ import {makeStyleSheetFromTheme, changeOpacity} from '@utils/theme';
-            color: theme.centerChannelColor,
+            color: changeOpacity(theme.centerChannelColor, 0.56),
+            fontSize: 15,
+            fontFamily: 'OpenSans',
```
Should we use `typography`? (and on the next style too)

larkox reviewed (approved):
Awesome! Just one small nit.

enahum commented on app/components/autocomplete/at_mention_group/at_mention_group.tsx:
```diff
@@ -81,8 +88,12 @@ const GroupMentionItem = ({
+            <View
+                style={style.rowInfo}
+            >
+                <Text style={style.rowUsername}>{`@${completeHandle} - `}</Text>
+                <Text style={style.rowFullname}>{`${completeHandle}`}</Text>
```
Why show it duplicate with and without the `@` sign?

shazm replied:
Whoops - fixed to use name + display name.

shazm commented on app/components/autocomplete/at_mention/at_mention.tsx:
```diff
@@ -287,16 +287,16 @@ const AtMention = ({
-        if (useGroupMentions) {
-            getGroupsForAutocomplete(serverUrl, channelId || '').then((res) => {
+        if (useGroupMentions && matchTerm && matchTerm !== '') {
+            getGroupsForAutocomplete(serverUrl, matchTerm || '').then((res) => {
```
Yup - will tidy up, but also if we are handling the team / channel constraints - will update once we know.

shazm commented on app/actions/remote/groups.ts:
```diff
@@ -4,13 +4,13 @@
     }
 
-    return client.getAllGroupsAssociatedToChannel(channelId, true);
+    return client.getGroups(query);
```
I assumed its because we need to get all groups, so I'm not sure how we are to handle the team / channel constraints alongside custom groups. Will circle back,

shazm commented:
<img width="564" alt="Screen Shot 2022-05-18 at 11 24 26 am" src="https://user-images.githubusercontent.com/456511/168945397-8ffaf50a-d1d5-4c5d-98f8-4417a8299239.png">

Styles updated for `Display name @name`

matthewbirtch commented:
> Styles updated for `Display name @name`

@shazm can you just update the following:
- color of the icon so that it matches what we do for `@channel` and `@all` autocomplete? 
- Also, it looks like the `@name` for the group is smaller. To be consistent with other autocomplete items, we should keep the font size consistent for all, so can you update that so it's the same font size as the display name (still keep the opacity you have through)

Here's what it should look like:
<img src="https://user-images.githubusercontent.com/2040554/169061636-d5afa451-6426-4db6-852d-0d34748856aa.png" width="414">

shazm commented:
Tidied up -
<img width="1041" alt="Screen Shot 2022-05-19 at 8 05 23 am" src="https://user-images.githubusercontent.com/456511/169163639-3d84bac0-bed0-4211-952b-0370baf7c9c9.png">

shazm commented:
@enahum - logic has been updated; tested against webapp behaviour.

- If there is a channel constraint; we only show groups in that channel and search/filter. 
- If there is a team constraint; we only show groups in that team and then search/filter.
- No constraints? Fetch / search on all groups.

enahum commented on app/components/autocomplete/at_mention/index.ts:
```diff
@@ -19,28 +20,44 @@ import type {WithDatabaseArgs} from '@typings/database/database';
 type OwnProps = {channelId?: string}
 const enhanced = withObservables([], ({database, channelId}: WithDatabaseArgs & OwnProps) => {
     const currentUser = observeCurrentUser(database);
+    const currentTeam = observeCurrentTeam(database);
+    const currentTeamId = observeCurrentTeamId(database);
```
Quick review, will check the rest tomorrow.

Two things:
1. With the team you have access to the teamId so no reason why querying this two.

Now the big thing, the team is not really the current team, it should be the team of the channel and fallback to the current team for DM/GM.

Why?
Autocomplete can be opened also from a thread, for example I can search for a post, saved post, recent mention etc.. and open the thread directly from there and they can be from any team.

shazm replied:
Oo yup very good point - will get it from channel instead and then revert to current team.

larkox commented on app/components/autocomplete/at_mention/at_mention.tsx:
```diff
@@ -353,16 +380,37 @@ const AtMention = ({
             setGroups(emptyGroupList);
         }
-    }, [channelId, useGroupMentions]);
+    }, [matchTerm, useGroupMentions]);
```
Shouldn't we depend on `isChannelConstrained` and `isTeamConstrained` too? Probably not a big deal, since matchTerm will change often enough.

shazm replied:
Yup - plus channel / team constraints are set in the system console by admins so would be pretty rare to change often enough to cause issues here.

larkox commented on app/components/autocomplete/at_mention/index.ts:
```diff
@@ -26,21 +28,48 @@ const enhanced = withObservables([], ({database, channelId}: WithDatabaseArgs &
         const currentChannel = observeChannel(database, channelId);
+        team = currentChannel.pipe(switchMap((c) => {
+            if (c && c.teamId) {
+                return c.team.observe() as Observable<TeamModel>;
```
Why do we need the `as`?

enahum replied:
because model.anotherModel.observe() returns `any` as its type

larkox replied:
It shouldn't be. The type of `c.team` is a `Relation<TeamModel>`, and a `Relation<T>`, the `observe()` function is of type `Observable<T|null>`. So... we shouldn't need or be doing the casting.

shazm replied:
`model.observe()` returns `Model | null`, but our `observeModel()` methods return `Model | undefined` - and I think I was getting type errors because of the difference between `null` and `undefined` further down in the code.

I have cast the model type there as if the `channel.teamId` is available, there should be a related team we can observe. 

I'm not sure, but a better way might be to pipe/map the observable and see if the team actually exists, and return the correct `null` or `undefined` behaviour as expected further down? 0/5 on which way might be better.

larkox commented on app/components/autocomplete/at_mention_group/at_mention_group.tsx:
```diff
@@ -66,8 +76,8 @@ const GroupMentionItem = ({
     const completeMention = useCallback(() => {
-        onPress(completeHandle);
-    }, [onPress, completeHandle]);
+        onPress(displayName);
```
Just confirming, do we want to autocomplete with the display name? Maybe I am missing something, but shouldn't we autocomplete with the name?

enahum replied:
this definitely needs to use the `name` instead of the `displayName`

shazm replied:
Whoops - I think I was testing something in there, will change.

larkox reviewed (approved):
LGTM, just a few small concerns, but should not be blocking.

shazm commented on app/components/autocomplete/at_mention/index.ts:
```diff
@@ -26,21 +28,48 @@ const enhanced = withObservables([], ({database, channelId}: WithDatabaseArgs &
         const currentChannel = observeChannel(database, channelId);
+        team = currentChannel.pipe(switchMap((c) => {
+            if (c && c.teamId) {
+                const mapTeam = (t: TeamModel | null) => (t ? t.observe() : of$(undefined));
+                return c.team.observe().pipe(switchMap(mapTeam));
```
@enahum / @larkox - I've changed it so I don't have to type cast it anymore; does that work better?

enahum replied:
you could actually use `c?.teamId ? observerTeam(database, c.teamId) : observeCurrentTeam(database)`

shazm replied:
This causes issues further down when I'm passing though the model, where it expects it to be `Model|undefined`, and those are inferred from the `Observable<Model...` so I'd have to do a lot of changing :(

enahum commented on app/components/autocomplete/at_mention_group/at_mention_group.tsx:
```diff
@@ -66,8 +76,8 @@ const GroupMentionItem = ({
-        onPress(completeHandle);
-    }, [onPress, completeHandle]);
+        onPress(name);
+    }, [onPress, displayName]);
```
```suggestion
    }, [onPress, name]);
```

enahum commented on app/actions/remote/groups.ts:
```diff
@@ -4,13 +4,32 @@
 import NetworkManager from '@managers/network_manager';
 
-export const getGroupsForAutocomplete = async (serverUrl: string, channelId: string) => {
+export const getGroupsForChannel = async (serverUrl: string, channelId: string) => {
```
rename to 
```suggestion
export const fetchGroupsForChannel = async (serverUrl: string, channelId: string) => {
```

enahum commented on app/actions/remote/groups.ts:
```diff
@@ -4,13 +4,32 @@
+};
 
-    return client.getAllGroupsAssociatedToChannel(channelId, true);
+export const getGroupsForTeam = async (serverUrl: string, teamId: string) => {
```
```suggestion
export const fetchGroupsForTeam = async (serverUrl: string, teamId: string) => {
```

enahum commented on app/actions/remote/groups.ts:
```diff
@@ -4,13 +4,32 @@
+    }
+};
+
+export const getGroupsForAutocomplete = async (serverUrl: string, query: string) => {
```
```suggestion
export const fetchGroupsForAutocomplete = async (serverUrl: string, query: string) => {
```

enahum commented on app/actions/remote/groups.ts:
```diff
@@ -4,13 +4,32 @@
+        return client.getAllGroupsAssociatedToChannel(channelId);
     } catch (error) {
-        return [];
+        return undefined;
```
Why ubdefined and not {error}, same for the others.. including the one returning an empty array

shazm replied:
Mostly to make the methods simpler where the fetch method is being called; the `fetch...` actions return either `Promise<Model>` or `undefined`, so i'm able to do things like `const data = await fetchSomeThing(); data?.myThing.map(...`.

For `autoComplete`, it plugs back into an array of sections to display the `@` mentions; it's much simpler to pass back an empty `[]` - we don't necessarily mind too much in this specific place that groups fetching may have failed.

There are three other chunks of work (group data handling, and then group push notification handling, post mention highlighting) where I'll be expanding the `catch(e){ ... }` to check what the error is, and retry, and potentially check the local db, or log the error and exit without asking the calling method to handle the error.

shazm replied:
For groups - we can pass the `{error}` to the method calling it, but there isn't much we  plan on doing with the `error` anyway - unless we specifically want to show to the user in the `@` mention list that the autocomplete for groups may have failed.

shazm replied:
Discussed with @enahum - will return `{error}` to stay consistent with other actions, and handle on the calling method 👍
